src/main/java/org/candlepin/subscriptions/capacity/CapacityReconciliationController.java

```diff
@@ -0,0 +1,115 @@
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ * Red Hat trademarks are not licensed under GPLv3. No permission is
+ * granted to use or replicate Red Hat trademarks that are incorporated
+ * in this software or its documentation.
+ */
+package org.candlepin.subscriptions.capacity;
+
+import static org.candlepin.subscriptions.db.model.SubscriptionCapacity.from;
+
+import io.micrometer.core.instrument.Counter;
+import io.micrometer.core.instrument.MeterRegistry;
+import java.util.*;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import lombok.extern.slf4j.Slf4j;
+import org.candlepin.subscriptions.capacity.files.ProductWhitelist;
+import org.candlepin.subscriptions.db.OfferingRepository;
+import org.candlepin.subscriptions.db.SubscriptionCapacityRepository;
+import org.candlepin.subscriptions.db.model.*;
+import org.springframework.stereotype.Component;
+
+@Component
+@Slf4j
+public class CapacityReconciliationController {
+
+  private final OfferingRepository offeringRepository;
+  private final ProductWhitelist productWhitelist;
+  private final CapacityProductExtractor productExtractor;
+  private final SubscriptionCapacityRepository subscriptionCapacityRepository;
+  private final Counter capacityRecordsCreated;
+  private final Counter capacityRecordsUpdated;
+  private final Counter capacityRecordsDeleted;
+
+  public CapacityReconciliationController(
+      OfferingRepository offeringRepository,
+      ProductWhitelist productWhitelist,
+      CapacityProductExtractor productExtractor,
+      SubscriptionCapacityRepository subscriptionCapacityRepository,
+      MeterRegistry meterRegistry) {
+    this.offeringRepository = offeringRepository;
+    this.productWhitelist = productWhitelist;
+    this.productExtractor = productExtractor;
+    this.subscriptionCapacityRepository = subscriptionCapacityRepository;
+    capacityRecordsCreated = meterRegistry.counter("rhsm-subscriptions.capacity.records_created");
+    capacityRecordsUpdated = meterRegistry.counter("rhsm-subscriptions.capacity.records_updated");
+    capacityRecordsDeleted = meterRegistry.counter("rhsm-subscriptions.capacity.records_deleted");
+  }
+
+  public void reconcileCapacityForSubscription(Subscription subscription) {
+
+    Collection<SubscriptionCapacity> newCapacities = mapSubscriptionToCapacities(subscription);
+    reconcileSubscriptionCapacities(
+        newCapacities, subscription.getSubscriptionId(), subscription.getSku());
+  }
+
+  private Collection<SubscriptionCapacity> mapSubscriptionToCapacities(Subscription subscription) {
+
+    Offering offering = offeringRepository.getById(subscription.getSku());
+    Set<String> products =
+        productExtractor.getProducts(
+            offering.getProductIds().stream().map(Object::toString).collect(Collectors.toSet()));
```

Comment:
Something seems a bit "extra" with this. That is:

1. offering already has a non-null set of Integers of productIds.
2. `productExtractor#getProducts` only takes a collection of Strings.
3. So this will convert the `Set<Integer>` into `Set<String>` and pass it to `getProducts`
4. What that method will do to those strings is make sure they're non-null and a valid integer and turn them back into Integers
5. It proceeds grabbing the products as necessary.

With some minor additions to CapacityProductExtractor, it is able to serve clients that have either a collection of Strings or a collection of Integers, like so:

```java
// In CapacityProductExtractor, getProducts splits into two methods:

  public Set<String> getProducts(Collection<String> productIds) {
    return getProductsByIds(
        productIds.stream()
            .map(CapacityProductExtractor::parseIntSkipUnparseable)
            .filter(Objects::nonNull));
  }

  public Set<String> getProductsByIds(Stream<Integer> productIds) {
    Set<String> products =
        productIds
            .map(engProductIdToSwatchProductIdsMap::get)
            .filter(Objects::nonNull)
            .flatMap(Set::stream)
            .collect(Collectors.toSet());

    if (setIsInvalid(products)) {
      // Kick out the RHEL products since it's implicit with the RHEL-included product being there.
      products = products.stream().filter(matchesRhel().negate()).collect(Collectors.toSet());
    }

    return products;
  }
```

Then, to use it, the above code becomes:

```java
Set<String> products = productExtractor.getProductsByIds(offering.getProductIds().stream());
```

Is it *worth* doing the above. One could argue that some clarity is added and redundancy reduced. Another would argue that CapacityProductExtractor is tested and works, so why change it.

Reply:
I'm fine with all the above suggestions. If the class-based projection suggestion starts to look like a time sink, then feel free to defer/descope.

Reply:
The change that @redsaz suggested makes sense to me and I made that update. 

As for using class based projections, I can't help feel that it's slightly over engineering, specifically for this use case. I am not against adding the projections but it does involve adding separate classes for the different types of projections we would want and that feels like something that needs to be looked at across the project instead of a single scenario. That way, we could have a list of the most common projections used that are more domain specific and can be reused. Also we are fetching a single offering which isn't usually that much of a concern in terms of size (imo that's the motivation for using projections, correct me if I am wrong). That's my 2c but I am open to changing my mind.

Reply:
In regards to "something that needs to be looked at across the project instead of a single scenario", I agree we should look at it across the project eventually, but if we chose to use projections we'd probably want to work that in incrementally instead.

Using projections (or even just simplifying some of the named query methods...I think you can do something like `Stream<String> findOffering_SkuBySku(String sku)` which would still cut down on the amount of data we're loading from the database.  If fetching a single offering doesn't do a ton of unnecessary eager fetching on other tables, I'm fine with leaving it the way it is and then revisiting projections if slowness/memory consumption becomes a problem.

Reply:
+1 to starting somewhere. I meant that we need to talk as a group about the possible projections that make sense fro the domain perspective. 

Looking into what the table design in like for offerings, looks like there is a join on sku between offering and sku_oid and the productIds are fetched eagerly. Still not a lot of data afaik, a sku has upto 10 engIds on an average. 

That being said, I will go ahead and add a ProductID projection, as it seems pretty straightforward and is an optimization without a lot of cost.

Reply:
@lindseyburnett after looking closely I realized that I do need the full offering entity because if you look at line 84, I do `products.stream()
        .map(product -> from(subscription, offering, product))
        .collect(Collectors.toList());` which basically builds subscription capacity like this

`.serviceLevel(offering.getServiceLevel())
        .usage(offering.getUsage())
        .sku(offering.getSku())
        .physicalSockets(offering.getPhysicalSockets())
        .virtualSockets(offering.getVirtualSockets())
        .virtualCores(offering.getVirtualCores())
        .physicalCores(offering.getPhysicalCores())`